gptfixer/gpt.c

```diff
@@ -0,0 +1,837 @@
+    if (starting_lba != 0x1) {
+        errx(EXIT_INVALID_MBR, "Bad protective MBR: does not start at first block");
+    }
+
+    uint32_t const expected_size_in_lba =
+        (uint32_t)(size_in_sectors - 1 > UINT32_MAX ? UINT32_MAX : size_in_sectors - 1);
+    if (size_in_lba != expected_size_in_lba) {
+        warnx("Bad protective MBR: does not cover whole disk: size in LBA is "
+              "%" PRIu32 ", expected %" PRIu32,
+              size_in_lba, expected_size_in_lba);
+        memcpy(mbr->partition_record[0].size_in_lba, &expected_size_in_lba, 4);
+    }
+    return changed;
+}
+
+static bool unused_entry(const struct GPTPartitionEntry *const entry) {
+    const char zero[sizeof(entry->PartitionTypeGUID)] = {0};
+    return memcmp(zero, &entry->PartitionTypeGUID, sizeof(zero)) == 0;
+}
+
+#define for_each_gpt_entry(i, size, count, gpt, entry)                                                                 \
+    if ((gpt)->header.SizeOfPartitionEntry < sizeof(struct GPTPartitionEntry) ||                                       \
+        (gpt)->header.SizeOfPartitionEntry % _Alignof(struct GPTPartitionEntry))                                       \
+        abort();                                                                                                       \
+    else                                                                                                               \
+        for (uint32_t i = 0, size = (gpt)->header.SizeOfPartitionEntry, count = (gpt)->used_entries; i < count; ++i)   \
+            for (const struct GPTPartitionEntry *entry =                                                               \
+                     (const struct GPTPartitionEntry *)((const char *)(gpt)->entries + i * size);                      \
+                 entry != NULL; entry = NULL)
+
+static void fixup_gpt(struct GPT *gpt, uint64_t bytes_required) {
+    uint64_t partition_entry_bytes = (uint64_t)gpt->used_entries * (uint64_t)gpt->header.SizeOfPartitionEntry;
+    assert(partition_entry_bytes <= UINT_MAX);
+    gpt->header.PartitionEntryArrayCRC32 =
+        (uint32_t)crc32(0UL, (const unsigned char *)gpt->entries, (unsigned int)partition_entry_bytes);
+    gpt->header.HeaderCRC32 = (uint32_t)crc32(0UL, (const unsigned char *)&gpt->header, sizeof(gpt->header));
+    memset((char *)gpt->entries + partition_entry_bytes, 0, bytes_required - partition_entry_bytes);
+}
+
+static struct GPT *gpt_adjust_sectors(uint64_t sector_offset, uint32_t sector_size, uint32_t old_sector_size,
+                                      uint64_t sectors, struct GPT *gpt, uint64_t *allocated) {
+    assert(sectors >= 6);
+    assert(valid_sector_size(old_sector_size));
+    assert(valid_sector_size(sector_size));
+    assert(sector_offset == 1 || sector_offset == sectors - 1);
+
+    uint64_t const size = sector_size * sectors;
+    struct GPT *new_gpt = aligned_alloc(sector_size, sector_size);
```

Comment:
A generic comment to all aligned_alloc calls: do you actually need those buffers to be sector-aligned? I don't think you use any API that requires such alignment.

Reply:
Originally the code used direct I/O, which does require alignment.  I switched to buffered I/O later because `O_SYNC` actually provided the semantics I need and because I might need to read 512 bytes from a 4096-byte-sector device.